torch/testing/_internal/opinfo/definitions/signal.py

```diff
@@ -0,0 +1,232 @@
+        sample_inputs_func=sample_inputs_func,
+        error_inputs_func=error_inputs_func,
+        supports_out=False,
+        supports_autograd=False,
+        skips=(
+            # TODO: same as this?
+            # https://github.com/pytorch/pytorch/issues/81774
+            # also see: arange, new_full
+            # fails to match any schemas despite working in the interpreter
+            DecorateInfo(
+                unittest.expectedFailure,
+                "TestOperatorSignatures",
+                "test_get_torch_func_signature_exhaustive",
+            ),
+            # fails to match any schemas despite working in the interpreter
+            DecorateInfo(
+                unittest.expectedFailure, "TestJit", "test_variant_consistency_jit"
+            ),
+            # skip these tests since we have non tensor input
+            DecorateInfo(
+                unittest.skip("Skipped!"), "TestCommon", "test_noncontiguous_samples"
+            ),
+            DecorateInfo(
+                unittest.skip("Skipped!"),
+                "TestCommon",
+                "test_variant_consistency_eager",
+            ),
+            DecorateInfo(unittest.skip("Skipped!"), "TestMathBits", "test_conj_view"),
+            DecorateInfo(
+                unittest.skip("Skipped!"), "TestMathBits", "test_neg_conj_view"
+            ),
+            DecorateInfo(unittest.skip("Skipped!"), "TestMathBits", "test_neg_view"),
+            DecorateInfo(
+                unittest.skip("Skipped!"),
+                "TestVmapOperatorsOpInfo",
+                "test_vmap_exhaustive",
+            ),
+            DecorateInfo(
+                unittest.skip("Skipped!"),
+                "TestVmapOperatorsOpInfo",
+                "test_op_has_batch_rule",
+            ),
+            DecorateInfo(
+                unittest.expectedFailure,
+                "TestSchemaCheckModeOpInfo",
+                "test_schema_correctness",
+                dtypes=[torch.float16],
+                device_type="cpu",
+            ),
+            DecorateInfo(
+                unittest.expectedFailure,
+                "TestDecomp",
+                "test_comprehensive",
+                dtypes=[torch.float16],
+                device_type="cpu",
+            ),
+            DecorateInfo(
+                unittest.expectedFailure,
+                "TestMeta",
+                "test_dispatch_meta",
+                dtypes=[torch.float16],
+                device_type="cpu",
+            ),
+            DecorateInfo(
+                unittest.expectedFailure,
+                "TestMeta",
+                "test_meta",
+                dtypes=[torch.float16],
+                device_type="cpu",
+            ),
+            *skips,
+        ),
+    )
+
+
+op_db: List[OpInfo] = [
+    make_signal_windows_opinfo(
+        name="signal.windows.cosine",
+        ref=make_signal_windows_ref(scipy.signal.windows.cosine)
```

Comment:
The test structures overall look great, but instead of creating the refs like this I suggest creating a reference function with the same signature as torch that manipulates the arguments to be compatible with SciPy, possibly performing additional conversions as necessary. There are some examples of this pattern in `common_methods_invocations.py`:

https://github.com/pytorch/pytorch/blob/5ffe24fca43b05216d55fa73a7d2629248315a30/torch/testing/_internal/common_methods_invocations.py#L2119

https://github.com/pytorch/pytorch/blob/5ffe24fca43b05216d55fa73a7d2629248315a30/torch/testing/_internal/common_methods_invocations.py#L7576

They're a little tricky to find because that file is very large and so many things are called "references." We'd probably choose more unique names given the chance.

Reply:
(In particular, I'm a little concerned about just disregarding the `dtype` parameter)

Reply:
Hmm... interesting. 

Now regarding the `dtype`, you'd like to also use the NumPy dtype argument? If we don't ignore that `dtype` it fails, so I guess we would have to work around that somehow.

```python
_______ TestCommonCPU.test_numpy_ref_signal_windows_gaussian_cpu_float64 _______
Traceback (most recent call last):
  File "/Users/agaona/github.com/pytorch/test/test_ops.py", line 162, in test_numpy_ref
    self.compare_with_reference(
  File "/Users/agaona/github.com/pytorch/torch/testing/_internal/common_utils.py", line 2352, in compare_with_reference
    expected = ref_fn(n_inp, *n_args, **n_kwargs)
  File "/Users/agaona/github.com/pytorch/torch/testing/_internal/opinfo/definitions/signal.py", line 117, in _fn
    return fn(*args, **kwargs)
TypeError: gaussian() got an unexpected keyword argument 'dtype'
```

Reply:
I would find the corresponding NumPy dtype using

https://github.com/pytorch/pytorch/blob/ce48df9e938ac208cf018545517344c6a6debab2/torch/testing/_internal/common_utils.py#L1040

and then cast the SciPy result to that dtype. This is imperfect because it won't perform the computation in that dtype (as best I can tell there's no way to request a dtype from windows like `scipy.signal.cosine()`) but it's the closest match for the PyTorch behavior, I think.

Reply:
Indeed, there's no way to perform the computation with a given `numpy.dtype`. What you're mentioning it seems that we have to do it here:
https://github.com/pytorch/pytorch/blob/0337f0ad473ffc298a30e603050d2df9d0073428/torch/testing/_internal/common_utils.py#L2365-L2366
But we need to check if `dtype` is present in `kwargs` to cast it appropriately, if not, then we leave it as it is; to avoid impacting other tests that do not pass `dtype` as a keyword arg. I don't think there's another way as we need access to the output of the ref function itself. What do you think?

Reply:
Hmm... I might be lying, I think we can do it in `make_signal_windows_ref` introduced in this PR, in the `signal.py` file.

Reply:
Mystery solved, the conversion happens here:
https://github.com/pytorch/pytorch/blob/0337f0ad473ffc298a30e603050d2df9d0073428/torch/testing/_internal/common_utils.py#L2361

Reply:
> The test structures overall look great, but instead of creating the refs like this I suggest creating a reference function with the same signature as torch that manipulates the arguments to be compatible with SciPy, possibly performing additional conversions as necessary.

@mruberry doing that implies that we'd have to create a reference function for each window function, not all of them follow the same signature, namely, `gaussian` and `exponential`. Others, do have the same exact signature, e.g., `cosine`, `hamming`, `hann`, etc.

Reply:
Why can't you use the same signature as we currently have with  `args` and `kwargs` and then simply perform the relevant casts to the output?

Reply:
As you suggested here: https://github.com/pytorch/pytorch/pull/85599#discussion_r992908369

Reply:
I would just create a reference function for each one since that'd be clearer to me, but whatever you like.

Casting to the correct NumPy dtype afterwards seems correct. A dtype of "None" should acquire the default floating point type, get its corresponding NumPy dtype, and cast to that, however. PyTorch's default floating point type is float32, but NumPy's default floating point type is (as mentioned above) float64, so we want to account for that discrepancy.

Reply:
> Why can't you use the same signature as we currently have with `args` and `kwargs` and then simply perform the relevant casts to the output?

@lezcano I might've been not very clear. I was referring to Mike's comment about creating a signature function that matches our torch implementation for the reference we're creating; not exactly the casting problem, which as you're saying we're doing with `astype()`. I believe Mike was saying to create a function for each window as it's been implemented in the example he provided. 

The solution you and I did is more general and I like it better, but it is true it's a little bit out of the standard.

> I would just create a reference function for each one since that'd be clearer to me, but whatever you like.
> 
> Casting to the correct NumPy dtype afterwards seems correct. A dtype of "None" should acquire the default floating point type, get its corresponding NumPy dtype, and cast to that, however. PyTorch's default floating point type is float32, but NumPy's default floating point type is (as mentioned above) float64, so we want to account for that discrepancy.

If I were to create a single function for all the references, I think I cannot comply with your request of having a torch-like signature, like the examples you provided [here](https://github.com/pytorch/pytorch/pull/85599#discussion_r992763771). It's a trade-off.

What do you guys think, Mario and Mike?

Reply:
Ok, let's leave it as it is. Mike gave us leeway to choose if we use one function per window or just a generic one.

Reply:
Either way is fine